proposals/p2687.md

```diff
@@ -0,0 +1,298 @@
+# Termination algorithm for impl selection
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2687)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Non-type arguments](#non-type-arguments)
+    -   [Proof of termination](#proof-of-termination)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Measure complexity using type tree depth](#measure-complexity-using-type-tree-depth)
+    -   [Consider each type parameter in an `impl` declaration separately](#consider-each-type-parameter-in-an-impl-declaration-separately)
+    -   [Consider types in the interface being implemented as distinct](#consider-types-in-the-interface-being-implemented-as-distinct)
+    -   [Require some count to decrease](#require-some-count-to-decrease)
+
+<!-- tocstop -->
+
+## Abstract
+
+This proposal replaces the termination algorithm for `impl` selection. The
+previous algorithm relied a recursion limit, which is counter to
+[our goal for predictability](/docs/design/generics/goals.md#predictability).
+
+The replacement is to terminate if any `impl` declaration depends transitively
+on itself with a "strict superset" type as one of its parameters.
```

Comment:
This sounds like it's a static property of the `impl` declaration itself, rather than a property of a particular query. I wonder if something like, "terminate if any `impl` lookup performed while considering an `impl` declaration depends transitively on the same `impl` declaration with a "strict superset" type as one of its parameters" would be clearer. But I guess there's a limit on how precise we should be in the Abstract.